client/cypress/integration/visualizations/word_cloud_spec.js

```diff
@@ -0,0 +1,83 @@
+    createQuery({ query: SQL }).then(({ id }) => {
+      cy.visit(`queries/${id}/source`);
+      cy.getByTestId('ExecuteButton').click();
+    });
+  });
+
+  it('creates visualization with automatic word frequencies', () => {
+    const visualizationName = 'Word Cloud (auto)';
+
+    cy.getByTestId('NewVisualization').click();
+    cy.getByTestId('VisualizationType').click();
+    cy.getByTestId('VisualizationType.WORD_CLOUD').click();
+    cy.getByTestId('VisualizationName').clear().type(visualizationName);
+
+    cy.getByTestId('WordCloud.WordsColumn').click();
+    cy.getByTestId('WordCloud.WordsColumn.a').click();
+
+    cy.getByTestId('VisualizationPreview').find('svg text').should('have.length', 11);
+
+    cy.getByTestId('EditVisualizationDialog').contains('button', 'Save').click();
+    cy.getByTestId('QueryPageVisualizationTabs').contains('li', visualizationName).should('exist');
+  });
+
+  it('creates visualization with word frequencies from another column', () => {
+    const visualizationName = 'Word Cloud (frequencies)';
+
+    cy.getByTestId('NewVisualization').click();
+    cy.getByTestId('VisualizationType').click();
+    cy.getByTestId('VisualizationType.WORD_CLOUD').click();
+    cy.getByTestId('VisualizationName').clear().type(visualizationName);
+
+    cy.getByTestId('WordCloud.WordsColumn').click();
+    cy.getByTestId('WordCloud.WordsColumn.b').click();
+
+    cy.getByTestId('WordCloud.FrequenciesColumn').click();
+    cy.getByTestId('WordCloud.FrequenciesColumn.c').click();
+
+    cy.getByTestId('VisualizationPreview').find('svg text').should('have.length', 5);
+
+    cy.getByTestId('EditVisualizationDialog').contains('button', 'Save').click();
+    cy.getByTestId('QueryPageVisualizationTabs').contains('li', visualizationName).should('exist');
+  });
+
+  it('creates visualization with word length and frequencies limits', () => {
+    const visualizationName = 'Word Cloud (filters)';
+
+    cy.getByTestId('NewVisualization').click();
+    cy.getByTestId('VisualizationType').click();
+    cy.getByTestId('VisualizationType.WORD_CLOUD').click();
+    cy.getByTestId('VisualizationName').clear().type(visualizationName);
+
+    cy.getByTestId('WordCloud.WordsColumn').click();
+    cy.getByTestId('WordCloud.WordsColumn.b').click();
+
+    cy.getByTestId('WordCloud.FrequenciesColumn').click();
+    cy.getByTestId('WordCloud.FrequenciesColumn.c').click();
+
+    cy.getByTestId('WordCloud.WordLengthLimit.Min').clear().type('4');
+    cy.getByTestId('WordCloud.WordLengthLimit.Max').clear().type('5');
+    cy.getByTestId('WordCloud.WordCountLimit.Min').clear().type('1');
+    cy.getByTestId('WordCloud.WordCountLimit.Max').clear().type('3');
+
+    cy.getByTestId('VisualizationPreview').find('svg text').should('have.length', 2);
+
+    cy.getByTestId('EditVisualizationDialog').contains('button', 'Save').click();
+    cy.getByTestId('QueryPageVisualizationTabs').contains('li', visualizationName).should('exist');
+  });
```

Comment:
Tests are great. A few things I would do different.
1. Swap some code blocks with `cy.clickThrough()` unless you think it's unnecessary.
2. Extract the `vis name -> save -> tab should exist` to a single test - bc it's testing a feature independent of the test's focus (which is the wordcloud outcome based on config). Also, might not even be a wordcloud feature but a visualization feature so might need to be extracted further.
3. Testing word length is good, but since size, color and position are also factors we'd like to protect from regression, I suggest a percy snapshot instead.

@kravets-levko wdyt?

Reply:
@ranbena Updated test according to your suggestions. Can you please take another look?